docs/architecture/adr-012-multi-version.md

```diff
@@ -0,0 +1,129 @@
+Core, making it solely a matter of Rust crate API evolution. In the long term,
+a single version of the tendermint-rs libraries should be able to support all
+versions of CometBFT Core protocols that are relevant to the community.
+The developers of CometBFT protocols should take versioning practices into use
+for future revisions that support backward compatibility on the source code level.
+
+## Decision
+
+The tendermint-rs library API will be modified to provide support for
+multiple versions of CometBFT Core protocol, at this moment being 0.34 and
+0.37.
+
+### tendermint-proto
+
+The generated Rust files providing bindings for Tendermint protobuf defitions
+will be emitted in two side-by-side modules, `tendermint::v0_34` and
+`tendermint::v0_37`. All names under the latter module are also reexported under
+the `tendermint` module, providing a low-change migration path for code bases
+that used previous versions of the crate, and the "default" import paths for
+those applications that only need to target the latest version.
+
+### tendermint
+
+The domain types are largely unchanged, except where additional fields are
+needed to support version 0.37. Some new types are added as required by the
+new version. Conversions to and from Protobuf types defined in
+`tendermint-proto` are provided for both `tendermint::v0_34` and
+`tendermint::v0_37` generated types, except where the types are new to 0.37.
+
+### tendermint-rpc
+
+A compatibility mode parameter is introduced for HTTP and WebSocket clients,
+with the following type:
+
+```rust
+pub enum CompatMode {
+    /// Use the v0.37 version of the protocol.
+    V0_37,
+    /// Use v0.34 version of the protocol.
+    V0_34,
+}
+```
+
+The mode parameter is used to dynamically select the encoding used by a client.
+By default, the clients use the latest protocol (designated by the associated
+const `CompatMode::LATEST`), but it's possible to specify
+the compatibility mode at construction using the newly introduced configuration
+API. In the HTTP client, it's also possible to switch the mode for a client
+that has been already connected; this is useful for dynamic version discovery
+using the `status` endpoint.
+
+The `Client` trait is extended to support methods new to CometBFT 0.37,
+which are emulated using older protocol endpoints when `V0_34` compatibility
+mode is selected. There is no need to deprecate any methods. All data types
+in the public client API are the domain types of `tendermint`, so the
+difference in encoding is confined to the crate internals.
+
+### tendermint-abci
+
+This crate is not actively supported and we should only make minimal effort
+to make it work with the proposed changes in a single chosen version of the
+protocol.
+
+## Open questions
+
+* Should we migrate tendermint-abci to 0.37, freeze it at 0.34 and declare
+  end of life, or commit to the same level of multi-version support as in
+  other crates?
```

Comment:
As discussed during the sync review, we've decided to migrate it to 0.37 in order to give its eventual customers a chance to update their apps for 0.37. We will eventually deprecate `tendermint-abci` in favor of [`tower-abci`](https://github.com/penumbra-zone/tower-abci) once the latter adds support for 0.37.

Let's update the ADR with this answer?

Reply:
OK. I'll add a negative consequence then: the existing `tendermint-abci` users who wish to remain on 0.34 will be effectively stranded on the previous release (0.29.x currently) of the crate.

Reply:
Good point!